.github/workflows/release.yml

```diff
@@ -6,28 +6,113 @@ on:
+      version: ${{ steps.ldflags.outputs.version }}
+      tree-state: ${{ steps.ldflags.outputs.tree-state }}
+    steps:
+      - name: Checkout
+        id: checkout
+        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # v3
+        with:
+          fetch-depth: 0
+
+      - name: Generate Ldflags
+        id: ldflags
+        run: |
+          echo "commit-date=$(git log --date=iso8601-strict -1 --pretty=%ct)" >> "$GITHUB_OUTPUT"
+          echo "commit=$GITHUB_SHA" >> "$GITHUB_OUTPUT"
+          echo "version=$(git describe --tags --always --dirty | cut -c2-)" >> "$GITHUB_OUTPUT"
+          echo "tree-state=$(if git diff --quiet; then echo "clean"; else echo "dirty"; fi)" >> "$GITHUB_OUTPUT"
+
   release:
+    permissions:
+      contents: write # needed to write releases
+      id-token: write # needed for keyless signing
     runs-on: ubuntu-latest
+    needs: args
+    outputs:
+      hashes: ${{ steps.hash.outputs.hashes }}
     steps:
-      - uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # v3.5.0
+      - name: Set Go Version
+        uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # v3.5.0
         with:
           go-version: 1.19
           check-latest: true
 
-      - uses: sigstore/cosign-installer@9becc617647dfa20ae7b1151972e9b3a2c338a2b # v2.8.1
-      - uses: anchore/sbom-action/download-syft@06e109483e6aa305a2b2395eabae554e51530e1d # v0.13.1
+      - name: Install Cosign
+        uses: sigstore/cosign-installer@9becc617647dfa20ae7b1151972e9b3a2c338a2b # v2.8.1
 
-      - uses: actions/checkout@755da8c3cf115ac066823e79a1e1788f8940201b # v3.2.0
+      - name: Install Syft
+        uses: anchore/sbom-action/download-syft@06e109483e6aa305a2b2395eabae554e51530e1d # v0.13.1
 
-      - uses: goreleaser/goreleaser-action@8f67e590f2d095516493f017008adc464e63adb1 # v4.1.0
-        with:
-          install-only: true
+      - name: Checkout
+        uses: actions/checkout@755da8c3cf115ac066823e79a1e1788f8940201b # v3.2.0
+
+      # - name: Set LDFLAGS
+      #   id: ldflags
+      #   run: |
+      #     echo "GO_FLAGS=( echo "-buildid= -X sigs.k8s.io/release-utils/version.gitVersion=${{needs.args.outputs.version}} \
+      #     -X sigs.k8s.io/release-utils/version.gitCommit==${{needs.args.outputs.commit}} \
+      #     -X sigs.k8s.io/release-utils/version.gitTreeState==${{needs.args.outputs.tree-state}} \
+      #     -X sigs.k8s.io/release-utils/version.buildDate==${{needs.args.outputs.commit-date}}" )" >> "$GITHUB_ENV"
 
-      - name: release
-        run: make release
+      - name: Run GoReleaser
+        id: run-goreleaser
+        uses: goreleaser/goreleaser-action@8f67e590f2d095516493f017008adc464e63adb1 # v4.1.0
+        with:
+          version: latest
+          args: release --rm-dist
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          LDFLAGS: echo "-buildid= -X sigs.k8s.io/release-utils/version.gitVersion=${{needs.args.outputs.version}} \
```

Comment:
if you use the generic generator, you can remove all the LDFLAG code you added for the Go builder. The generic generator does not use any of these flags, so you can keep the original code as-is. You only need the `Generate subject` step and the `provenance` job to be added to the workflow.

Reply:
I do think the ldflags do need to get set somewhere though.

It seems that helm-sigstore doesn't use `sigs.k8s.io/release-utils` and instead sets the version in `cmd/version.go`.
https://github.com/sigstore/helm-sigstore/blob/60bfeab071df7ac8d7405f8d98d3ba2b00488853/cmd/version.go#L29-L43

I think this is only set when built via the `Makefile` so you'd need to update the goreleaser.yml to include the right ldflags or set the `LDFLAGS` environment variable here.
https://github.com/sigstore/helm-sigstore/blob/60bfeab071df7ac8d7405f8d98d3ba2b00488853/Makefile#L49

I think simply replacing `sigs.k8s.io/release-utils/version` with `github.com/helm-sigstore/cmd` would work here.

Reply:
I though on generating the LDFLAGS dynamically in the workflow to be more clear where they are coming from. 

Is it a good practice to use by makefile or using this version? Or is it okay to keep it on workflow run?

Reply:
I think I'll try the same approach as https://github.com/sigstore/policy-controller/blob/main/.github/workflows/release.yaml to get the ldflags. 

I tried to generate it on github workflow to be clearer to the reader where it is being from but maybe it is not the best practice.